File: client/src/models/workspaces.ts
import { serializable } from "@/utils/serializable";
import { immerable } from "immer";
import { Variable } from "@/models/variables";
import { Environment } from "@/models/environments";

export class Workspace {
  [immerable] = true;

  @serializable("string", { required: true, format: "uuid" })
  id!: string;

  @serializable("string", { required: true, format: "identifier" })
  name!: string;

  @serializable("string")
  description?: string;

  @serializable("object", { factory: Environment })
  environment?: Environment[];

  @serializable("object", { factory: Variable })
  variables?: Variable[];

  constructor(object?: Partial<Workspace>) {
    Object.assign(this, object);
    this.variables = object?.variables?.map((v) => new Variable(v));
    this.environment = object?.environment?.map((e) => new Environment(e));
  }
}

File: agent/src/models/workspaces.rs
use serde::{ Deserialize, Serialize };

use crate::models::environments::Environment;
use crate::models::variables::Variable;

#[derive(Serialize, Deserialize)]
pub struct Workspace {
    pub id: String,
    pub name: String,
    pub description: String,
    pub environments: Vec<Environment>,
    pub variables: Vec<Variable>,
}

impl Default for Workspace {
    fn default() -> Self {
        Self {
            id: uuid::Uuid::new_v4().to_string(),
            name: String::new(),
            description: String::new(),
            environments: Vec::new(),
            variables: Vec::new(),
        }
    }
}
